notifications: Implement the notification aspect of the "Followed Topic" feature.

prakhar1144 opened the pull request:
#12309 

Add support for:
- [x] Followed Topic email notifications
- [x] Followed Topic push notifications
- [x] Followed Topic wildcard mention notifications
- [ ] Followed Topic desktop notifications
- [ ] Followed Topic audible notifications

<!-- Describe your pull request here.-->


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on templates/zerver/emails/missed_message.html:
```diff
@@ -33,6 +33,8 @@
     {% elif stream_email_notify %}
     {% trans %}You are receiving this because you have email notifications enabled for #{{ stream_name }}.{% endtrans %}<br />
+    {% elif followed_topic_email_notify %}
+    {% trans %}You are receiving this because you have email notifications enabled for the followed topic #{{ stream_name }} > {{ topic_name }}.{% endtrans %}<br />
```
@alya thoughts on this string? I think it could be just "... because you are following #{{ stream_name }} > {{topic_name}}.", or "because you have email notifications enabled for topics you follow."; I think the latter might be preferable, since it implies pretty clearly that you follow this topic anyway, with the extra length/clutter of spelling out the topic.

alya replied:
Yeah, I like the latter option.

prakhar1144 replied:
Updated the string.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -223,12 +225,16 @@ def get_recipient_info(
                 ),
                 user_profile_push_notifications=F("user_profile__enable_stream_push_notifications"),
                 user_profile_wildcard_mentions_notify=F("user_profile__wildcard_mentions_notify"),
+                followed_topic_email_notifications=F(
```
We might want to call these e.g. `user_profile_followed_topic_email_notifications`, for future-proofing in case in the future we offer different "followed topic" overrides for individual streams...

But I think there's no reason to do that preemptively, given that these variables names are entirely local to this function and the longer versions would be very cumbersome.

prakhar1144 replied:
> But I think there's no reason to do that preemptively, given that these variables names are entirely local to this function and the longer versions would be very cumbersome.

Didn't rename in favour of the above statement.

timabbott commented on zerver/lib/stream_subscription.py:
```diff
@@ -350,6 +350,7 @@ def get_subscriptions_for_send_message(
         | (Q(push_notifications=None) & Q(user_profile__enable_stream_push_notifications=True))
         | Q(email_notifications=True)
         | (Q(email_notifications=None) & Q(user_profile__enable_stream_email_notifications=True))
+        | Q(user_profile__enable_followed_topic_email_notifications=True)
```
So I think in a world where we set `enable_followed_topic_email_notifications` to true by default, this line will defeat the performance benefits of this function, documented in its docstring. 

I think the fix for that problem is that we need to instead include every user who actually follows this topic in who we fetch these data for. 

Possibly a good way to do that would be for `get_recipient_info` to be responsible for this.

prakhar1144 replied:
> this line will defeat the performance benefits of this function, documented in its docstring.

Agreed.

> we need to instead include every user who actually follows this topic in who we fetch these data for.

Makes sense.
I have updated the logic.

> Possibly a good way to do that would be for get_recipient_info to be responsible for this.

We can do this, but it will lead to an increase in query count by 1.
Let me know, if the current approach has any downsides, we can change this accordingly.

prakhar1144 commented:
@abhijeetbodas2001 @timabbott 

This is ready to review.

Changes in the recent push:
* Addressed the review comments. (Please check the respective review comments)
* Earlier the commit was incomplete -- completed the remaining work. 

Pending design discussions:
* Default value of the fields -- [CZO discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/notifications.20of.20Followed.20Topics/near/1572780)

abhijeetbodas2001 commented on zerver/models.py:
```diff
@@ -1588,6 +1588,13 @@ class UserBaseSettings(models.Model):
+    enable_followed_topic_email_notifications = models.BooleanField(default=False)
+    enable_followed_topic_push_notifications = models.BooleanField(default=False)
+    enable_followed_topic_audible_notifications = models.BooleanField(default=False)
+    enable_followed_topic_wildcard_mentions_notify = models.BooleanField(default=True)
```
Why not end this with `notifications` (instead of `notify`) like the other fields?

prakhar1144 replied:
I did this to keep it consistent with the existing:

```python
    # Stream notifications.
    enable_stream_desktop_notifications = models.BooleanField(default=False)
    enable_stream_email_notifications = models.BooleanField(default=False)
    enable_stream_push_notifications = models.BooleanField(default=False)
    enable_stream_audible_notifications = models.BooleanField(default=False)
    notification_sound = models.CharField(max_length=20, default="zulip")
    wildcard_mentions_notify = models.BooleanField(default=True)
```

timabbott replied:
Yeah, consistent is probably correct. It is a significantly different behavior, since it's whether wildcard mentions are eligible for notifications at all.

abhijeetbodas2001 commented on zerver/tests/test_users.py:
```diff
@@ -1979,6 +1980,36 @@ def test_stream_recipient_info(self) -> None:
         self.assertEqual(info.all_bot_user_ids, {normal_bot.id, service_bot.id})
 
+        # Now Hamlet follows the topic with 'followed_topic_email_notifications' global setting enabled.
+        hamlet.enable_followed_topic_email_notifications = True
+        hamlet.save()
```
We should instead use `do_change_user_setting` here instead of manually changing the DB.
(Not high priority, though also a ping that it would be worthwhile to split this test suite whenever you are free :))

prakhar1144 replied:
Hm, used `do_change_user_setting`.

abhijeetbodas2001 reviewed (commented):
This looks good, I left some comments below. One question I have is, if we're going ahead with implementing email notifications first completely, and then push notifications, when should we create a single migrations for all the settings (like this PR does currently), or independent migrations for all the settings?

prakhar1144 commented on zerver/lib/message.py:
```diff
@@ -156,6 +156,7 @@ class SendMessageRequest:
     pm_mention_email_disabled_user_ids: Set[int]
     stream_push_user_ids: Set[int]
     stream_email_user_ids: Set[int]
+    followed_topic_email_user_ids: Set[int]
```
Comment added. 
(Although it was making complete sense to me via only the variable name.)
Adding a comment to be more clear makes sense. 👍🏻

prakhar1144 commented:
@timabbott @abhijeetbodas2001 This is ready to review.

[**Update 1.0**](https://github.com/zulip/zulip/pull/25640#issuecomment-1555626477)

Updates in the recent push:
* Addressed the review comments. (Please check the respective review comments)
* Fields, default values set to `true` -- [CZO Discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/notifications.20of.20Followed.20Topics/near/1575417)

[Comment](https://github.com/zulip/zulip/pull/25640#pullrequestreview-1436165465) -- [CZO discussion](https://chat.zulip.org/#narrow/stream/3-backend/topic/Follow.20Topics/near/1576024):
> One question I have is, if we're going ahead with implementing email notifications first completely, and then push notifications, when should we create a single migrations for all the settings (like this PR does currently), or independent migrations for all the settings?

prakhar1144 commented:
Update: Implemented the `push notifications` part.

prakhar1144 commented:
@zulipbot add "mentor review"

abhijeetbodas2001 commented on zerver/tests/test_event_queue.py:
```diff
@@ -760,6 +760,33 @@ def test_stream_email_and_push_notify_unmuted_topic_and_unmuted_stream(
+        do_change_user_setting(
+            self.user_profile, "enable_followed_topic_push_notifications", False, acting_user=None
+        )
+        do_set_user_topic_visibility_policy(
```
Similar to this, I think we should also add a test in the previous commit to verify that the email setting when disabled does not send notifications.

prakhar1144 replied:
Organized the new tests in this file in a better way. (While working on the wildcard mention part, I observed the better way in which these new tests can be organized).

[Newly added tests](https://github.com/zulip/zulip/pull/25640/commits/c2e10004496fbf0dcccf210f4c1ae159f9bfe56e#diff-b7ecab40a23cd87e0acf58e7c8220330787bb4f7bb59e8ba601ae648ffdc968aR763)

abhijeetbodas2001 reviewed (commented):
I left some minor comments below, and posted a question in https://chat.zulip.org/#narrow/stream/378-api-design/topic/allowed.20visibility.20policies.
Apart from that, this looks good to me!

prakhar1144 commented on zerver/openapi/zulip.yaml:
```diff
@@ -9753,6 +9753,15 @@ paths:
+          in: query
+          description: |
+            Enable email notifications for followed topic messages.
+
```
Done.

prakhar1144 commented:
@abhijeetbodas2001 @timabbott 

This is ready to review.

Updates in the recent push:
* Addressed the review comment. (Please check the respective review comment.)
* Implemented the `wildcard mention notifications` part.

prakhar1144 commented:
@zulipbot add "buddy review"

timabbott commented on zerver/tests/test_event_queue.py:
```diff
@@ -760,6 +760,56 @@ def test_stream_email_and_push_notify_unmuted_topic_and_unmuted_stream(
+                user_id=self.user_profile.id,
+                followed_topic_email_notify=False,
+                already_notified={"email_notified": False, "push_notified": False},
+            )
```
I think merging this test with the previous one would make the difference between the test setups clearer.

timabbott commented on zerver/lib/notification_data.py:
```diff
@@ -186,6 +205,8 @@ def get_email_notification_trigger(self, acting_user_id: int, idle: bool) -> Opt
+        elif self.followed_topic_wildcard_mention_email_notify:
+            return NotificationTriggers.FOLLOWED_TOPIC_WILDCARD_MENTION
         elif self.wildcard_mention_email_notify:
             return NotificationTriggers.WILDCARD_MENTION
```
This is an interesting corner case about precedence that we should discuss in #design. (It's easy to change, so not a blocker for this PR). I think it's possible that in the case where `wildcard_mention_email_notify` was already true, even though the "followed topic" case is more fine-grained, it may not be worth calling out that you follow the topic in the description of why you got the email, since you are getting it because of the wildcard mention regardless in your configuration?

timabbott commented on zerver/migrations/0455_alter_scheduledmessagenotificationemail_trigger.py:
```diff
@@ -0,0 +1,26 @@
+                    ("wildcard_mentioned", "Wildcard mention"),
+                    ("stream_email_notify", "Stream notifications enabled"),
+                    ("followed_topic_email_notify", "Followed Topic notifications enabled"),
+                    ("followed_topic_wildcard_mentioned", "Followed Topic wildcard mention"),
```
This migration should be dropped in favor of adding the new fields to this existing migration:
`zerver/migrations/0331_scheduledmessagenotificationemail.py`

As noted before, this is not actually a database change, it's just a change to Django's Python code understanding of what values are possible, and it's a distraction to treat it as an independent migration.

timabbott replied:
Pushed this in my revisions to this PR.

timabbott commented:
I renumbered the migrations and made the changes proposed in https://github.com/zulip/zulip/pull/25640#discussion_r1228830138; with those changes, this looks good to me, marked to merge once CI passes.

* https://github.com/zulip/zulip/pull/25640#discussion_r1228829182 is a follow-up design conversation
* https://github.com/zulip/zulip/pull/25640#discussion_r1228823795 might be worth a follow-up to make it easier to see why those tests are correct.

Thanks for all the work on this @prakhar1144 and @abhijeetbodas2001! This is huge progress towards once of our most important projects goals for the summer.
